File: togo/internal/audit/audit_test.go
package audit

import (
	"testing"

	"github.com/nikeee/cappu/internal/packages"
)

// stubSource returns canned advisories keyed by coordinate.
type stubSource struct {
	byCoord map[packages.CoordinateString][]Advisory
}

func (s stubSource) Name() string { return "stub" }

func (s stubSource) FindVulnerabilities(coords []packages.Coordinates) (map[packages.CoordinateString][]Advisory, error) {
	return s.byCoord, nil
}

func TestAuditPackagesReportSortedWorstFirst(t *testing.T) {
	src := stubSource{byCoord: map[packages.CoordinateString][]Advisory{
		"org.low:low:1":   {{ID: "L", Severity: SeverityLow}},
		"org.crit:crit:1": {{ID: "C", Severity: SeverityCritical}, {ID: "M", Severity: SeverityModerate}},
	}}
	coords := []packages.Coordinates{
		coord("org.low:low:1"),
		coord("org.clean:clean:1"), // no advisories
		coord("org.crit:crit:1"),
		coord("org.crit:crit:1"), // duplicate, deduped
	}
	report, err := AuditPackages(coords, src)
	if err != nil {
		t.Fatal(err)
	}
	if report.Scanned != 3 { // duplicate deduped, clean still counts as scanned
		t.Errorf("scanned = %d, want 3", report.Scanned)
	}
	if len(report.Vulnerable) != 2 {
		t.Fatalf("vulnerable = %d, want 2", len(report.Vulnerable))
	}
	// worst-first: crit (has a critical) before low
	if report.Vulnerable[0].Coordinates.String() != "org.crit:crit:1" {
		t.Errorf("first vulnerable = %s, want org.crit:crit:1", report.Vulnerable[0].Coordinates.String())
	}
	if report.Counts.Critical != 1 || report.Counts.Moderate != 1 || report.Counts.Low != 1 {
		t.Errorf("counts = %+v", report.Counts)
	}
	if report.Counts.Total() != 3 {
		t.Errorf("total = %d, want 3", report.Counts.Total())
	}
}

func TestAuditPackagesNoneVulnerable(t *testing.T) {
	report, err := AuditPackages([]packages.Coordinates{coord("org.a:a:1")}, stubSource{})
	if err != nil {
		t.Fatal(err)
	}
	if report.Scanned != 1 || len(report.Vulnerable) != 0 {
		t.Errorf("report = %+v", report)
	}
}

File: src/audit/audit.test.ts
import { test } from "node:test";

import { expect } from "expect";

import {
  type Coordinates,
  coordinatesToString,
  type CoordinateString,
  toCoordinates,
} from "../packages/index.ts";
import { auditPackages } from "./audit.ts";
import { type Advisory, type AdvisoryId, type AuditSource } from "./types.ts";

const advisory = (id: string, severity: Advisory["severity"]): Advisory => ({
  id: id as AdvisoryId,
  aliases: [],
  summary: id,
  severity,
  fixedVersions: [],
  url: `https://osv.dev/vulnerability/${id}`,
});

// A canned source: maps coordinate strings to advisories.
function source(map: Record<string, Advisory[]>): AuditSource {
  return {
    name: "fake",
    findVulnerabilities: (coordinates: readonly Coordinates[]) => {
      const result = new Map<CoordinateString, Advisory[]>();
      for (const c of coordinates) {
        const a = map[coordinatesToString(c)];
        if (a) result.set(coordinatesToString(c), a);
      }
      return Promise.resolve(result);
    },
  };
}

test("auditPackages tallies counts, dedupes coordinates and sorts worst-first", async () => {
  const coords: Coordinates[] = [
    toCoordinates("org", "low", "1"),
    toCoordinates("org", "crit", "1"),
    toCoordinates("org", "crit", "1"), // duplicate (e.g. main + test)
    toCoordinates("org", "clean", "1"),
  ];
  const report = await auditPackages(
    coords,
    source({
      "org:low:1": [advisory("L1", "low")],
      "org:crit:1": [advisory("C1", "critical"), advisory("M1", "moderate")],
    }),
  );

  expect(report.scanned).toBe(3); // deduped: crit counted once
  expect(report.counts).toEqual({ critical: 1, high: 0, moderate: 1, low: 1, unknown: 0 });
  // crit (worst = critical) before low
  expect(report.vulnerable.map(p => p.coordinates.artifactId)).toEqual(["crit", "low"]);
});

test("auditPackages reports nothing for a clean set", async () => {
  const report = await auditPackages([toCoordinates("org", "clean", "1")], source({}));
  expect(report.vulnerable).toEqual([]);
  expect(report.scanned).toBe(1);
});
